front/lib/api/wrappers.ts

```diff
@@ -128,3 +130,112 @@ export function withSessionAuthenticationForWorkspace<T>(
     opts
   );
 }
+
+/**
+ * This function is a wrapper for Public API routes that require authentication for a workspace.
+ * It must be used on all routes that require workspace authentication (prefix: /v1/w/[wId/]).
+ *
+ * opts.allowUserOutsideCurrentWorkspace allows the handler to be called even if the key is not a
+ * associated with the workspace. This is useful for routes that share data across workspaces (eg apps
+ * runs).
+ *
+ * @param handler
+ * @param opts
+ * @returns
+ */
+export function withPublicApiAuthentication<T, U extends boolean>(
+  handler: (
+    req: NextApiRequest,
+    res: NextApiResponse<WithAPIErrorResponse<T>>,
+    auth: Authenticator,
+    keyAuth: U extends true ? Authenticator : null
+  ) => Promise<void> | void,
+  opts: {
+    isStreaming?: boolean;
+    allowUserOutsideCurrentWorkspace?: U;
+  } = {}
+) {
+  const { allowUserOutsideCurrentWorkspace, isStreaming } = opts;
+
+  return withLogging(
+    async (
+      req: NextApiRequest,
+      res: NextApiResponse<WithAPIErrorResponse<T>>
+    ) => {
+      const { wId } = req.query;
+      if (typeof wId !== "string" || !wId) {
+        return apiError(req, res, {
+          status_code: 404,
+          api_error: {
+            type: "workspace_not_found",
+            message: "The workspace was not found.",
+          },
+        });
+      }
+
+      const keyRes = await getAPIKey(req);
+      if (keyRes.isErr()) {
+        return apiError(req, res, keyRes.error);
+      }
+
+      const keyAndWorkspaceAuth = await Authenticator.fromKey(
+        keyRes.value,
+        wId,
+        getGroupIdsFromHeaders(req.headers)
+      );
+      const { keyAuth } = keyAndWorkspaceAuth;
+      let { workspaceAuth } = keyAndWorkspaceAuth;
+
+      const owner = workspaceAuth.workspace();
+      const plan = workspaceAuth.plan();
+      if (!owner || !plan) {
+        return apiError(req, res, {
+          status_code: 404,
+          api_error: {
+            type: "workspace_not_found",
+            message: "The workspace was not found.",
+          },
+        });
+      }
+
+      // Authenticator created from the a key has the builder role if the key is associated with the workspace.
+      if (!workspaceAuth.isBuilder() && !allowUserOutsideCurrentWorkspace) {
+        return apiError(req, res, {
+          status_code: 401,
+          api_error: {
+            type: "workspace_auth_error",
+            message: "Only users of the workspace can access this route.",
+          },
+        });
+      }
+
+      // /!\ This is reserved for internal use!
+      // If the header "x-api-user-email" is present and valid,
+      // exchange the workspace auth for a user auth.
```

Comment:
nit: indentation is a bit off. We could mention here that it only works for system key

Reply:
Fixed.